pandas/core/groupby/generic.py

```diff
@@ -129,1358 +128,1348 @@ def pinner(cls):
+            raise AssertionError("Number of keys must be 1")
 
-            result = concat(results).sort_index()
+        axis = self.axis
+        obj = self._obj_with_exclusions
+
+        result = OrderedDict()
+        if axis != obj._info_axis_number:
+            try:
+                for name, data in self:
+                    result[name] = self._try_cast(func(data, *args, **kwargs), data)
+            except Exception:
+                return self._aggregate_item_by_item(func, *args, **kwargs)
         else:
-            result = Series()
+            for name in self.indices:
+                try:
+                    data = self.get_group(name, obj=obj)
+                    result[name] = self._try_cast(func(data, *args, **kwargs), data)
+                except Exception:
+                    wrapper = lambda x: func(x, *args, **kwargs)
+                    result[name] = data.apply(wrapper, axis=axis)
 
-        # we will only try to coerce the result type if
-        # we have a numeric dtype, as these are *always* udfs
-        # the cython take a different path (and casting)
-        dtype = self._selected_obj.dtype
-        if is_numeric_dtype(dtype):
-            result = maybe_downcast_to_dtype(result, dtype)
+        return self._wrap_generic_output(result, obj)
 
-        result.name = self._selected_obj.name
-        result.index = self._selected_obj.index
-        return result
+    def _aggregate_item_by_item(self, func, *args, **kwargs):
+        # only for axis==0
 
-    def _transform_fast(self, func, func_nm):
-        """
-        fast version of transform, only applicable to
-        builtin/cythonizable functions
-        """
-        if isinstance(func, str):
-            func = getattr(self, func)
+        obj = self._obj_with_exclusions
+        result = OrderedDict()
+        cannot_agg = []
+        errors = None
+        for item in obj:
+            data = obj[item]
+            colg = SeriesGroupBy(data, selection=item, grouper=self.grouper)
 
-        ids, _, ngroup = self.grouper.group_info
-        cast = self._transform_should_cast(func_nm)
-        out = algorithms.take_1d(func()._values, ids)
-        if cast:
-            out = self._try_cast(out, self.obj)
-        return Series(out, index=self.obj.index, name=self.obj.name)
+            try:
+                cast = self._transform_should_cast(func)
 
-    def filter(self, func, dropna=True, *args, **kwargs):  # noqa
-        """
-        Return a copy of a Series excluding elements from groups that
-        do not satisfy the boolean criterion specified by func.
+                result[item] = colg.aggregate(func, *args, **kwargs)
+                if cast:
+                    result[item] = self._try_cast(result[item], data)
 
-        Parameters
-        ----------
-        func : function
-            To apply to each group. Should return True or False.
-        dropna : Drop groups that do not pass the filter. True by default;
```

Comment:
aggregate was not a copy paste because there was a def in both classes. However, the current master for DataFrameGroupBy just calls super, so that function body was simply replaced with what was previously in NDFrameGroupBy